src/sprite_data.cpp

```diff
@@ -333,9 +408,76 @@ const uint8 *ImageData::GetRecoloured(GradientShift shift, const Recolouring &re
 		}
 	}
 
-	ptr = result.get();
-	this->recoloured.emplace(map_key, std::move(result));
-	return ptr;
+	return result;
+}
+
+/**
+ * Scale this image to a different size.
+ * @param factor Factor by which to scale.
+ * @return The scaled instance.
+ */
+const ImageData *ImageData::Scale(float factor) const
+{
+	constexpr float EPSILON = 0.01;  ///< Threshold for float comparisons.
+	if (fabs(factor - 1.0f) < EPSILON) return this;
```

Comment:
Why don't you use the actual needed sprite pixel size versus the supplied sprite pixel size(s)?

These real-valued factors tend to introduce glitches due to rounding.